tool/randomizer.lua

```diff
@@ -184,14 +141,66 @@ function randomizer.randomize()
         local newlevel = workingset[newlevelindex]
         table.insert(visited, newlevel)
         table.remove(workingset, newlevelindex)
-        for _, nowvalid in pairs(dependants[queue[1].difficulty][newlevel] or {}) do
+        for _, nowvalid in pairs(dependants[difficulty][newlevel] or {}) do
             if not table.contains(visited, nowvalid) and not table.contains(workingset, nowvalid) then
                 table.insert(workingset, nowvalid)
             end
         end
-        table.insert(levelsequence, newlevel - 2)
+        table.insert(sequence, newlevel - 2)
+    end
+    table.insert(sequence, 31 - 2)
+    return sequence
+end
+
+function randomizer.randseed()
+    local difficulty = difficulty
+    local seed = os.time() .. math.floor(os.clock()*10000)
+    local sequence = randomizer.createsequence(seed, difficulty)
+    queue = {queue[1], {difficulty = difficulty, seed = seed, sequence = sequence, seedtype = randomizer.SEEDTYPE.RANDOMSEED}}
+end
+
+function randomizer.setseed(seed)
+    local difficulty = difficulty
+    local sequence = randomizer.createsequence(seed, difficulty)
+    if queue[2].seedtype == randomizer.SEEDTYPE.KEEPSEED or queue[2].seedtype == randomizer.SEEDTYPE.RANDOMSEED then
+        queue[2] = {difficulty = difficulty, seed = seed, sequence = sequence, seedtype = randomizer.SEEDTYPE.SETSEED}
+    else
+        table.insert(queue, {difficulty = difficulty, seed = seed, sequence = sequence, seedtype = randomizer.SEEDTYPE.SETSEED})
     end
-    table.insert(levelsequence, 31 - 2)
+end
+
+function randomizer.setsequence(sequence, sequencestr)
+    if queue[2].seedtype == randomizer.SEEDTYPE.KEEPSEED or queue[2].seedtype == randomizer.SEEDTYPE.RANDOMSEED then
+        queue[2] = {sequence = sequence, seedtype = randomizer.SEEDTYPE.SETSEQUENCE, sequencestr = sequencestr}
+    else
+        table.insert(queue,{sequence = sequence, seedtype = randomizer.SEEDTYPE.SETSEQUENCE, sequencestr = sequencestr})
+    end
+end
+
+local function nextlevel()
+    if levelindex <= #levelsequence then
+        tas:set_level(levelsequence[levelindex])
+    end
+    levelindex = levelindex + 1
+end
```

Comment:
How do you handle keep-sequence / keep-seed? Isn't this just always incrementing the level-list?

Reply:
`levelindex` was not created by me, but all it does is it remembers how many buttons were pressed. This is relevant for the count of buttons in `randomizer.hudlines()`, other than that it's used only in `nextlevel()` and changed when appropriate in `randomizer.run()`.

I actually left this untouched, not sure why it  says i added it

Reply:
Oh yeah, nevermind. It's in `nextlevel`, I thought it was in the onreset function.